packages/pubsub/src/publisher.js

```diff
@@ -0,0 +1,179 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*!
+ * @module pubsub/publisher
+ */
+
+'use strict';
+
+var arrify = require('arrify');
+var common = require('@google-cloud/common');
+var extend = require('extend');
+var is = require('is');
+
+/**
+ *
+ */
+function Publisher(topic, options) {
+  this.topic = topic;
+  this.api = topic.api;
+
+  this.inventory = {
+    queued: [],
+    inFlight: []
+  };
+
+  Object.defineProperty(this.inventory, 'queueSize', {
+    get: function() {
+      return this.queued.reduce(function(size, message) {
+        return size + message.data.length;
+      }, 0);
+    }
+  });
+
+  this.settings = extend(true, {
+    batching: {
+      maxBytes: Infinity,
+      maxMessages: Infinity,
+      maxMilliseconds: 1000
+    }
+  }, options);
+
+  this.publishHandle_ = null;
+}
+
+/**
+ *
+ */
+Publisher.prototype.cancelPublish_ = function() {
+  clearTimeout(this.publishHandle_);
+  this.publishHandle_ = null;
+};
+
+/**
+ *
+ */
+Publisher.prototype.prepareMessages_ = function() {
+  var size = 0;
+  var messages = [];
+
+  var message;
+
+  while (message = this.inventory.queued.shift()) {
+    var isUnderSizeLimit = (size + message.data.length) <=
+      this.settings.batching.maxBytes;
+
+    var isUnderMessageLimit = (messages.length + 1) <=
+      this.settings.batching.maxMessages;
+
+    if (!isUnderSizeLimit || !isUnderMessageLimit) {
+      this.inventory.queued.unshift(message);
+      break;
+    }
+
+    messages.push(message);
+  }
+
+  return messages;
+};
+
+/**
+ *
+ */
+Publisher.prototype.publish = function(data, attrs, callback) {
+  if (!(data instanceof Buffer)) {
+    throw new Error('Data must be in the form of a Buffer.');
+  }
+
+  if (is.fn(attrs)) {
+    callback = attrs;
+    attrs = {};
+  }
+
+  this.inventory.queued.push({
+    data: data,
+    attrs: attrs,
+    callback: callback
+  });
+
+  var reachedMaxMessages = this.inventory.queued.length >=
+    this.settings.batching.maxMessages;
+
+  var reachedMaxBytes = this.inventory.queueSize >=
+    this.settings.batching.maxBytes;
+
+  // publish immediately if we hit maxMessage or maxBytes
+  if (reachedMaxMessages || reachedMaxBytes) {
+    this.publish_();
+  } else if (!this.timeoutHandle_) {
+    this.schedulePublish_();
+  }
+};
+
+/**
+ *
+ */
+Publisher.prototype.publish_ = function() {
+  var self = this;
+
+  this.cancelPublish_();
+
+  var messages = this.prepareMessages_();
+
+  var reqOpts = {
+    topic: this.topic.name,
+    messages: messages.map(function(message) {
+      return {
+        data: message.data,
+        attributes: message.attrs
+      };
+    })
+  };
+
+  this.inventory.inFlight = messages;
```

Comment:
This part is a bit concerning.

If I am reading this code correctly, you basically have a single batch object that juggles its messages from an acceptance queue to an in flight queue. My concern: What happens if this process is not serial?

In other words, say you have a bunch of messages (`M1`) that start coming in at 12:00:00.000. At 12:00:01.000, enough time has elapsed and it is time to commit the batch. `this.inventory.inFlight` is set to the list of messages, which presumably gets cleared below (although I do not see it at quick glance), and you are off to the races.

The next batch, `M2`, starts at 12:00:01.000, and goes for a second like before. But wait, the response from `M1`'s in-flight status is not back yet. It turns out that Pub/Sub is a little slow or there is network latency or `{reason}` and the request is taking a second and a half.

Now at this point, you are overwriting the `M1` messages with the `M2` messages, which (if I read this correctly) means that when the response comes back, it will be comparing the message ID list for `M1` against the list of messages in `M2`. This is obviously a bad thing. :-)

Am I correct in my assessment, or did I miss something?